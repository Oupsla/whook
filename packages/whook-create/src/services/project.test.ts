import initProject from './project';
import YError from 'yerror';

describe('initProject', () => {
  const CWD = '/home/whoiam/projects/';
  const inquirer = { prompt: jest.fn() };
  const lock = { take: jest.fn(), release: jest.fn() };
  const ensureDir = jest.fn();
  const log = jest.fn();

  beforeEach(() => {
    lock.take.mockReset();
    inquirer.prompt.mockReset();
    lock.release.mockReset();
    ensureDir.mockReset();
    log.mockReset();
  });

  it('should work', async () => {
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      projectName: 'super-project',
    });
    inquirer.prompt.mockResolvedValueOnce({
      projectDirectory: '/home/whoiam/projects/yolo',
    });
    lock.release.mockResolvedValueOnce(undefined);
    ensureDir.mockResolvedValueOnce(undefined);

    const project = await initProject({
      inquirer,
      CWD,
      ensureDir,
      lock,
      log,
    });

    expect({
      project,
      inquirerPromptCalls: inquirer.prompt.mock.calls,
      lockTakeCalls: lock.take.mock.calls,
      lockReleaseCalls: lock.release.mock.calls,
      ensureDirCalls: ensureDir.mock.calls,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should fail with access problems', async () => {
    lock.take.mockResolvedValueOnce(undefined);
    lock.release.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      projectName: 'super-project',
    });
    inquirer.prompt.mockResolvedValueOnce({
      projectDirectory: '/home/whoiam/projects/yolo',
    });
    ensureDir.mockRejectedValueOnce(new YError('E_ACCESS'));

    try {
      await initProject({
        inquirer,
        CWD,
        ensureDir,
        lock,
        log,
      });
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        errorCode: err.code,
        errorParams: err.params,
        inquirerPromptCalls: inquirer.prompt.mock.calls,
        lockTakeCalls: lock.take.mock.calls,
        lockReleaseCalls: lock.release.mock.calls,
        ensureDirCalls: ensureDir.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => 'stack' !== type),
      }).toMatchSnapshot();
    }
  });
});
