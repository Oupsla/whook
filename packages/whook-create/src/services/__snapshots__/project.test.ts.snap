// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initProject should fail with access problems 1`] = `
Object {
  "ensureDirCalls": Array [
    Array [
      "/home/whoiam/projects/yolo",
    ],
  ],
  "errorCode": "E_PROJECT_DIR",
  "errorParams": Array [
    "/home/whoiam/projects/yolo",
  ],
  "inquirerPromptCalls": Array [
    Array [
      Array [
        Object {
          "default": "new-whook-project",
          "message": "What's this new project name",
          "name": "projectName",
        },
      ],
    ],
    Array [
      Array [
        Object {
          "default": "/home/whoiam/projects/super-project",
          "message": "Provide the project's directory",
          "name": "projectDirectory",
        },
      ],
    ],
  ],
  "lockReleaseCalls": Array [
    Array [
      "cli:input",
    ],
    Array [
      "cli:input",
    ],
  ],
  "lockTakeCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üèÅ - Initializing project...",
    ],
    Array [
      "error",
      "Cannot create the project's directory:",
      "/home/whoiam/projects/yolo",
    ],
  ],
}
`;

exports[`initProject should work 1`] = `
Object {
  "ensureDirCalls": Array [
    Array [
      "/home/whoiam/projects/yolo",
    ],
  ],
  "inquirerPromptCalls": Array [
    Array [
      Array [
        Object {
          "default": "new-whook-project",
          "message": "What's this new project name",
          "name": "projectName",
        },
      ],
    ],
    Array [
      Array [
        Object {
          "default": "/home/whoiam/projects/super-project",
          "message": "Provide the project's directory",
          "name": "projectDirectory",
        },
      ],
    ],
  ],
  "lockReleaseCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "lockTakeCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üèÅ - Initializing project...",
    ],
  ],
  "project": Object {
    "directory": "/home/whoiam/projects/yolo",
    "name": "super-project",
  },
}
`;
